rust/benches/file_backend.rs

```diff
@@ -201,44 +206,74 @@ fn file_backend_benchmark_matrix(c: &mut Criterion) {
+                    ));
+                    group.plot_config(plot_config.clone());
+                    for backend_fn in backend_open_fns() {
+                        file_backend_benchmark(
+                            &mut group, path, operation, access, chunk_size, threads, backend_fn,
+                        );
+                    }
                 }
             }
         }
     }
 }
 
-fn file_backend_benchmark<M: Measurement>(
-    g: &mut BenchmarkGroup<'_, M>,
+fn file_backend_benchmark(
+    g: &mut BenchmarkGroup<'_, WallTime>,
     path: &Path,
     operation: Operation,
     access: AccessPattern,
     chunk_size: usize,
+    threads: usize,
     backend_fn: BackendOpenFn,
 ) {
     let mut options = OpenOptions::new();
     options.create(true).read(true).write(true);
 
     let (backend, backend_name) = backend_fn(path, options.clone()).unwrap();
 
+    let mut completed_iterations = 0u64;
     g.throughput(Throughput::Bytes(chunk_size as u64));
     g.bench_with_input(
         BenchmarkId::from_parameter(backend_name),
         // these are passed though [criterion::black_box]
-        &(operation, access, chunk_size, backend),
-        |b, (operation, access, chunk_size, backend)| {
-            let mut data = vec![0; *chunk_size];
-            let mut iter = 0;
-            b.iter(|| {
-                let offset = access.offset(iter, *chunk_size);
-                operation.execute(backend.as_ref(), &mut data, offset, iter);
-                iter += 1;
+        &(operation, access, chunk_size, threads, backend),
+        |b, (operation, access, chunk_size, threads, backend)| {
+            let chunk_size = *chunk_size;
+            let threads = *threads;
+            b.iter_custom(|iterations| {
+                let start_toggle = &AtomicBool::new(false);
+                let duration = std::thread::scope(|s| {
+                    let mut handles = Vec::with_capacity(threads);
+                    for thread in 0..threads {
+                        handles.push(s.spawn(move || {
+                            let mut data = vec![0; chunk_size];
+                            while !start_toggle.load(Ordering::Acquire) {
+                                std::hint::spin_loop();
+                            }
+                            for iter in ((completed_iterations + thread as u64)
+                                ..(completed_iterations + iterations))
+                                .step_by(threads)
+                            {
+                                let offset = access.offset(iter, chunk_size);
+                                operation.execute(backend.deref(), &mut data, offset, iter);
+                            }
+                        }));
+                    }
+                    let start = Instant::now();
+                    start_toggle.store(true, Ordering::Release);
+
+                    for handle in handles {
+                        handle.join().unwrap();
+                    }
```

Comment:
Not sure if it matters, but instead of measuring the join time here, each thread could measure it's own time and then do a `AtomicU64::fetch_max`.

Reply:
Good point. I opted for last_end - first_start.